smolyay/smolyak.py

```diff
@@ -0,0 +1,156 @@
+import numpy
+
+
+class IndexGrid:
+    """Indexes of grid points for constructing a surrogate function.
+
+    Parameters
+    ----------
+    dimension : int
+        Number of independent variables.
+    exactness : int
+        Level of exactness of the approximation.
+    index_per_level : list
+        Number of indexes at each level.
+    """
+
+    def __init__(self, dimension, exactness, index_per_level):
+        self.dimension = dimension
+        self.exactness = exactness
+        self.index_per_level = index_per_level
+
+    @property
+    def dimension(self):
+        """Get dimension."""
+        return self._dimension
+
+    @dimension.setter
+    def dimension(self, value):
+        print('Set dimension')
+        self._dimension = value
+        self._needs_update = True
+
+    @property
+    def exactness(self):
+        """Get exactness."""
+        return self._exactness
+
+    @exactness.setter
+    def exactness(self, value):
+        print('Set exactness')
+        self._exactness = value
+        self._needs_update = True
+
+    @property
+    def index_per_level(self):
+        """Get index_per_level."""
+        return self._index_per_level
+
+    @index_per_level.setter
+    def index_per_level(self, value):
+        print('Set index_per_level')
+        self._index_per_level = value
+        self._needs_update = True
+
+    def _update(self):
+        """Update the indexes of grid points.
+
+        Generates the indexes for each level
+        depending on the index_per_level, and makes the
+        indexes of grid points.
```

Comment:
You should document any exceptions that you might raise (e.g., IndexError) here in a `Raises` section.